node/src/components/consensus/highway_core/state.rs

```diff
@@ -768,31 +862,76 @@ impl<C: Context> State<C> {
-            return Err(UnitError::FaultyCreator);
+        match (wunit.seq_num_panorama.get(creator), wunit.previous) {
+            (ObservationSeqNum::Faulty, _) => return Err(UnitError::FaultyCreator),
+            (ObservationSeqNum::None, Some(_)) | (ObservationSeqNum::Correct(_), None) => {
+                return Err(UnitError::PreviousUnit);
+            }
+            (ObservationSeqNum::Correct(seq_num), Some(_)) => {
+                if seq_num.checked_add(1) != Some(wunit.seq_number) {
+                    return Err(UnitError::SequenceNumber);
+                }
+            }
+            (ObservationSeqNum::None, None) if wunit.seq_number != 0 => {
+                return Err(UnitError::SequenceNumber);
+            }
+            _ => {}
+        }
+        if let Some(panorama) = maybe_panorama {
+            if panorama.len() != self.validator_count() {
+                return Err(UnitError::PanoramaLength(panorama.len()));
+            }
+            if panorama.get(creator).correct() != wunit.previous.as_ref() {
+                return Err(UnitError::PreviousUnit);
+            }
+            for ((vidx, obs), sn_obs) in panorama.enumerate().zip(&wunit.seq_num_panorama) {
+                match (obs, sn_obs) {
+                    (Observation::Correct(_), ObservationSeqNum::Correct(_))
+                    | (Observation::None, ObservationSeqNum::None)
+                    | (Observation::Faulty, ObservationSeqNum::Faulty) => {}
+                    (_, _) => return Err(UnitError::PanoramaMismatch(vidx)),
+                }
+            }
         }
         Ok(())
     }
 
     /// Returns an error if `swunit` is invalid. Must only be called once `pre_validate_unit`
     /// returned `Ok` and all dependencies have been added to the state.
-    pub(crate) fn validate_unit(&self, swunit: &SignedWireUnit<C>) -> Result<(), UnitError> {
+    pub(crate) fn validate_unit(
+        &self,
+        swunit: &SignedWireUnit<C>,
+        panorama: &Panorama<C>,
+    ) -> Result<(), UnitError> {
         let wunit = swunit.wire_unit();
         let creator = wunit.creator;
-        let panorama = &wunit.panorama;
         let timestamp = wunit.timestamp;
         panorama.validate(self)?;
         if panorama.iter_correct(self).any(|v| v.timestamp > timestamp)
             || wunit.timestamp < self.params.start_timestamp()
         {
             return Err(UnitError::Timestamps);
         }
+        for ((vidx, obs), seq_obs) in panorama.enumerate().zip(&wunit.seq_num_panorama) {
+            match (obs, seq_obs) {
+                (Observation::Faulty, ObservationSeqNum::Faulty)
+                | (Observation::None, ObservationSeqNum::None) => {}
+                (Observation::Correct(hash), ObservationSeqNum::Correct(seq_number))
+                    if self.unit(hash).seq_number == *seq_number => {}
+                (_, _) => return Err(UnitError::PanoramaMismatch(vidx)),
+            }
+        }
```

Comment:
Can we extract this to a method on `Panorama` type? Also, it seems like you do that in `pre_validate_unit` as well.

Reply:
It's not exactly the same, though, because in `validate_unit` we compare sequence numbers, and in `pre_validate_unit` we can't (because we don't necessarily have the unit with the cited hash yet).